janus_core/src/retries.rs

```diff
@@ -0,0 +1,315 @@
+//! Provides a simple interface for retrying fallible HTTP requests.
+
+use backoff::{future::retry, ExponentialBackoff};
+use futures::Future;
+use reqwest::StatusCode;
+use std::time::Duration;
+use tracing::{debug, warn};
+
+/// Traverse chain of source errors looking for an `std::io::Error`.
+fn find_io_error(original_error: &dyn std::error::Error) -> Option<&std::io::Error> {
+    let mut cause = original_error.source();
+    while let Some(err) = cause {
+        if let Some(typed) = err.downcast_ref() {
+            return Some(typed);
+        }
+        cause = err.source();
+    }
+
+    None
+}
+
+/// An [`ExponentialBackoff`] with parameters suitable for most HTTP requests. The parameters are
+/// copied from the parameters used in the GCP Go SDK[1].
+///
+/// AWS doesn't give us specific guidance on what intervals to use, but the GCP implementation cites
+/// AWS blog posts so the same parameters are probably fine for both, and most HTTP APIs for that
+/// matter.
+///
+/// [1]: https://github.com/googleapis/gax-go/blob/fbaf9882acf3297573f3a7cb832e54c7d8f40635/v2/call_option.go#L120
+pub fn http_request_exponential_backoff() -> ExponentialBackoff {
+    ExponentialBackoff {
+        initial_interval: Duration::from_secs(1),
+        max_interval: Duration::from_secs(30),
+        multiplier: 2.0,
+        max_elapsed_time: Some(Duration::from_secs(600)),
+        ..Default::default()
+    }
+}
+/// An [`ExponentialBackoff`] with parameters tuned for tests where we don't want to be retrying
+/// for 10 minutes.
+#[cfg(feature = "test-util")]
+pub fn test_http_request_exponential_backoff() -> ExponentialBackoff {
```

Comment:
This function is only used in this package's tests; move into the test module as a private function?

Reply:
(I'd also be OK with inlining this at points of call, or defining it in a const/static way somewhere -- this function returns a contant value so it may be better expressed as something other than a function)

Reply:
Good catch: the intention was to use this in `janus_client`'s tests so that they wouldn't possibly block for ten minutes.